### PR TITLE
Adding support for view interactions as well as input, file, and mult…

### PR DESCRIPTION
**Pull request by tsein-stripe**
Adding support for view interactions as well as input, file, and multi select block elements

### CODE REVIEW

**Comment by IRuslan** [comment]
@tsein-stripe I wonder why have you removed `omitempty` struct tag value from structs like `publishViewRequest`. Without this, `views.publish` requests fail with an empty hash (which is valid accordingly Slack API specs)?

**Comment by tsein-stripe** [comment]
> @tsein-stripe I wonder why have you removed `omitempty` struct tag value from structs like `publishViewRequest`. Without this, `views.publish` requests fail with an empty hash (which is valid accordingly Slack API specs)?

Thanks! I fixed that with my last commit.

**Comment by dcelasun** [comment]
OK I've managed to test this internally and it's looking good! The removal of `SingularBlockElement` required some code changes, but that's understandable. We don't use *all* of these new elements, but the ones we do worked correctly, so LGTM.

**Comment by tsein-stripe** [comment]
@kanata2 Do you have any concerns or PR feedback, or can this be merged and PR's into slack-go/slack?

File: block_conv.go [hunk 1]
```diff
@@ -65,11 +67,8 @@ func (b *Blocks) UnmarshalJSON(data []byte) error {
 		case "rich_text":
 			// for now ignore the (complex) content of rich_text blocks until we can fully support it
 			continue
-		case "file":
-			// for now ignore the file blocks until we can fully support it
-			continue
 		default:
-			return errors.New("unsupported block type")
+			block = &UnknownBlock{}
```
Comment (kanata2) [inline block_conv.go:68]: Any block type which may be introduced by Slack might want to be handled the same way as `rich_text`. (ignore until now)

File: views.go [hunk 2]
```diff
@@ -12,6 +12,10 @@ const (
 
 type ViewType string
 
+type ViewState struct {
+	Values map[string]map[string]BlockAction `json:"values"`
```
Comment (kanata2) [inline views.go:16]: 👍

**Review by kanata2 — COMMENTED** [review]
Sorry for slow response. Just added some comments.

File: block_conv.go (hunk 1 above)
Comment (tsein-stripe) [reply block_conv.go:68]: I can go either way here -- this was a change I brought in from
https://github.com/acaloiaro/slack/pull/1/files
after the following issue was filed:
https://github.com/slack-go/slack/issues/579

Skipping these elements also solves the issue of blowing up during unmarshalling new elements as Slack releases them -- but having the UnknownBlock element type increases transparency of when this happens, and keeps the unmarshalled form truer to the marshalled form (nothing gets omitted).

What do you think?

Comment (kanata2) [reply block_conv.go:68]:
```go
type UnknownBlock struct {
	Type     MessageBlockType `json:"type"`
	BlockID  string           `json:"block_id,omitempty"`
	Elements BlockElements    `json:"elements"`
}
```

I think new blocks are not necessarily defined the above properties and we might miss some  properties when marshal/unmarshal.

> but having the UnknownBlock element type increases transparency of when this happens

However this thought looks good to me.

**Comment by tsein-stripe** [comment]
@kanata2 Is this ready to be merged in and PR'd to the upstream slack-go repo?

**Comment by kanata2** [comment]
@tsein-stripe 
What do you do with this? future tasks?
https://github.com/kanata2/slack/pull/1#discussion_r383692634

**Comment by tsein-stripe** [comment]
@kanata2 What do you think about the last change I submitted? Certainly all blocks will at the very least have a type/ID. I also included rich text as a block which will get mapped to the placeholder "UnknownBlock" type.

Again, I prefer having the UnknownBlock placeholder here so that it's fully transparent to the user that something came through from a Slack API response which this framework doesn't yet have an impl for -- rather than skipping and omitting the element (which could be confusing)

**Comment by kanata2** [comment]
Looks good in terms of finding unsupported blocks without parsing failure.
You'll receive an error when you send json included marshalled UnknownBlock (e.g. `{"type": "rich_text", "block_id": "hoge"}`), so we need to handle this problem in next PR.

**Review by kanata2 — APPROVED** [review]

**Comment by tsein-stripe** [comment]
Thanks @kanata2 ! @dcelasun do you guys need anything else from me, or can you railroad these changes into the upstream repo? There should be no merge conflicts.

**Comment by tsein-stripe** [comment]
I figured out how to get this PR against the upstream repo -- I didn't know you could do that 🤦‍♂ 
https://github.com/slack-go/slack/pull/669

**Comment by kanata2** [comment]
close this PR. New one is [here](slack-go#669)!

**Pull request closed without merging**